main.go

```diff
@@ -616,6 +308,499 @@ func main() {
+	f, err := os.Open(cfg.rbacConfigPath)
+	if err != nil {
+		return nil, err
+	}
+
+	defer f.Close()
+
+	authorizer, err := rbac.Parse(f)
+	if err != nil {
+		return nil, err
+	}
+
+	return authorizer, nil
+}
+
+// Load tenant config information from tenants.yaml.
+func loadTenantConfigs(cfg *config, tCfg *tenantsConfig) {
+	tCfg.logger = logger.NewLogger(cfg.logLevel, cfg.logFormat, cfg.debug.name)
+
+	f, err := ioutil.ReadFile(cfg.tenantsConfigPath)
+	if err != nil {
+		stdlog.Fatalf("cannot read tenant configuration file from path %q: %v", cfg.tenantsConfigPath, err)
+	}
+
+	if err := yaml.Unmarshal(f, &tCfg); err != nil {
+		stdlog.Fatalf("unable to read tenant YAML: %v", err)
+	}
+
+	skip := level.Warn(log.With(tCfg.logger, "msg", "skipping invalid tenant"))
+
+	for _, t := range tCfg.Tenants {
+		err = loadOIDCConf(tCfg, t)
+		if err != nil {
+			skip.Log("tenant", t.Name, "err", err)
+		}
+
+		err = loadMTLSConf(tCfg, t)
+		if err != nil {
+			skip.Log("tenant", t.Name, "err", err)
+		}
+
+		err = newAuthorizer(tCfg, t)
+		if err != nil {
+			skip.Log("tenant", t.Name, "err", err)
+		}
+	}
+
+	// Load RBAC config information from rbac.yaml.
+	tCfg.authorizer, err = loadRBACConfig(*cfg)
+	if err != nil {
+		stdlog.Fatalf("unable to read RBAC YAML: %v", err)
+		return
+	}
+
+	tCfg.reg = prometheus.NewRegistry()
+
+	tCfg.reg.MustRegister(
+		version.NewCollector("observatorium"),
+		collectors.NewGoCollector(),
+		collectors.NewProcessCollector(collectors.ProcessCollectorOpts{}),
+	)
+
+	if cfg.middleware.rateLimiterAddress != "" {
+		ctx, cancel := context.WithTimeout(context.Background(), grpcDialTimeout)
+		defer cancel()
+
+		tCfg.rateLimitClient = ratelimit.NewClient(tCfg.reg)
+		if err := tCfg.rateLimitClient.Dial(ctx, cfg.middleware.rateLimiterAddress); err != nil {
+			stdlog.Fatal(err)
+		}
+	}
+
+	tCfg.ins = signalhttp.NewHandlerInstrumenter(tCfg.reg, []string{"group", "handler"})
+}
+
+// Create autorizer for a tenant.
+func newAuthorizer(tCfg *tenantsConfig, t *tenant) error {
+	skip := level.Warn(log.With(tCfg.logger, "msg", "skipping invalid tenant"))
+
+	if t.OPA != nil {
+		if t.OPA.URL != "" {
+			u, err := url.Parse(t.OPA.URL)
+			if err != nil {
+				skip.Log("tenant", t.Name, "err", fmt.Sprintf("failed to parse OPA URL: %v", err))
+
+				return err
+			}
+
+			t.OPA.authorizer = opa.NewRESTAuthorizer(u, opa.LoggerOption(log.With(tCfg.logger, "tenant", t.Name)))
+		} else {
+			a, err := opa.NewInProcessAuthorizer(t.OPA.Query, t.OPA.Paths, opa.LoggerOption(log.With(tCfg.logger, "tenant", t.Name)))
+			if err != nil {
+				skip.Log("tenant", t.Name, "err", fmt.Sprintf("failed to create in-process OPA authorizer: %v", err))
+
+				return err
+			}
+			t.OPA.authorizer = a
+		}
+	}
+
+	return nil
+}
+
+// Load MTLS auth information for a tenant.
+func loadMTLSConf(tCfg *tenantsConfig, t *tenant) error {
+	var err error
+
+	skip := level.Warn(log.With(tCfg.logger, "msg", "skipping invalid tenant"))
+
+	if t.MTLS != nil {
```

Comment:
ditto

Reply:
Same as above, I wonder if we can check that oncaller side - much simpler to read + non surprising